### PR TITLE
Fix FSDP2 crashing when tensors escape forward pass

### PR DESCRIPTION
**Pull request by arkadip-maitra**
Fixes #173709 

**Issue:**
FSDP2 crashes during backward when tensors escape the forward pass (e.g., via caching like k/v caching in attention) and the loss depends on those escaped tensors

**Cause:**
After `post_forward` the unsharded parameter storage is immediately freed. However, if a tensor that escaped the forward is later used to compute the loss, the backward pass tries to access the freed storage

**Fix:** 
Defer storage freeing when output tensors require gradients:

Added test

@wconstab Can you review this pls. Not sure if this is the optimal fix.

### CODE REVIEW

File: torch/distributed/fsdp/_fully_shard/_fsdp_param_group.py [hunk 1]
```diff
@@ -455,14 +461,18 @@ def post_forward(self, module: nn.Module, input: Any, output: Any):
         if not compiled_autograd_enabled():
             logger.debug("%s", self._with_fqn("FSDP::post_forward"))
         with record_function(self._with_fqn("FSDP::post_forward")):
+            # if any output requires grad defer storage freeing until post_backward
```
Comment (arkadip-maitra) [inline torch/distributed/fsdp/_fully_shard/_fsdp_param_group.py:465]: @wconstab also is this condition okay or should defering freeing storage be user controlled through the api param?

**Comment by weifengpy** [comment]
will take a look tonight

File: torch/distributed/fsdp/_fully_shard/_fsdp_param_group.py (hunk 1 above)
Comment (weifengpy) [reply torch/distributed/fsdp/_fully_shard/_fsdp_param_group.py:465]: for pretraining, model.parameters() all require grad, `defer_storage_free=True` becomes equivalent to `reshard_after_forwad=False`. see below script, the error would be gone with `reshard_after_forwad=False`

I actually think it's the cache's resonsibility to register hooks to call FSDPModule.unshard and .reshard. moving the discussion to the original post

```
"""
torchrun --nproc_per_node=2 test_fsdp2_sdpa_backward.py
"""

import os
import torch
import torch.nn as nn
import torch.distributed as dist
import torch.distributed.fsdp


class TestModule(nn.Module):
    def __init__(self):
        super().__init__()
        self.linear1 = nn.Linear(64, 64)
        self.linear2 = nn.Linear(64, 64)

    def forward(self, x):
        # Compute intermediate result and cache it
        intermediate = self.linear1(x)
        self.cached_intermediate = intermediate

        # Return a different output (that loss won't depend on)
        output = self.linear2(intermediate)
        return output


def test_fsdp2_backward_with_cached_inputs():
    """
    Test that FSDP2 works correctly when loss depends on cached intermediate
    tensors rather than module outputs.
    """
    dist.init_process_group("nccl")
    local_rank = int(os.environ.get("LOCAL_RANK", 0))
    torch.cuda.set_device(local_rank)
    module = TestModule().cuda()
    mp_policy = torch.distributed.fsdp.MixedPrecisionPolicy(
        param_dtype=torch.float32,
        reduce_dtype=torch.float32,
        output_dtype=torch.float32,
        cast_forward_inputs=True,
    )
    torch.distributed.fsdp.fully_shard(
        module=module,
        reshard_after_forward=False, # set it to True results in error
        mp_policy=mp_policy,
    )
    x = torch.randn(4, 64, requires_grad=True).cuda()
    output = module(x)
    loss = module.cached_intermediate.sum()
    loss.backward()
    dist.destroy_process_group()


if __name__ == "__main__":
    test_fsdp2_backward_with_cached_inputs()

```

Comment (arkadip-maitra) [reply torch/distributed/fsdp/_fully_shard/_fsdp_param_group.py:465]: ok thanks
